README.rst
================
sprockets-status
================
Simple application status handler for Tornado.

+-------------+--------------------------------------------------------+
| |GitHub|    | https://github.com/sprockets/sprockets-status          |
+-------------+--------------------------------------------------------+
| |PyPI|      | https://pypi.org/project/sprockets-status/             |
+-------------+--------------------------------------------------------+
| |Docs|      | https://pythonhosted.org/sprockets-status/             |
+-------------+--------------------------------------------------------+
| |Coveralls| | https://coveralls.io/github/sprockets/sprockets-status |
+-------------+--------------------------------------------------------+
| |Travis|    | https://travis-ci.org/sprockets/sprockets-status       |
+-------------+--------------------------------------------------------+

This library includes a very simple ``tornado.web.RequestHandler`` that
returns an applications status information.  You should install the
handler under the ``/status`` path for a very simple healthcheck that
verifies the operation of the IOLoop (at the very least).  This library
is the successor to the `sprockets.handlers.status`_ library.  We have
moved away from a deeply nested package structure based on namespace
packages because they are actually more trouble than they are worth.

Simplest Example
================
The simplest example is a JSON payload that contains directly configured
information.  The application name and version is passed through the
``URLSpec`` keyword arguments to the status handler.

.. code-block:: python

   from tornado import ioloop, web
   import sprockets_status.handlers

   def make_app(**settings):
       return web.Application([
           web.url('/status', sprockets_status.handlers.StatusHandler,
                   {'name': 'my-app-name', 'version': '1.1.1'}),
           # add your handlers here
       ], **settings)

   if __name__ == '__main__':
       app = make_app()
       iol = ioloop.IOLoop.current()
       try:
           app.listen(8888)
           iol.start()
       except KeyboardInterrupt:
           iol.stop()

Running this application and retrieving the ``/status`` resource returns
the following:

.. code-block:: http

   HTTP/1.1 200 OK
   Content-Length: 62
   Etag: "e7bca140bba5af0fdb7b9e4fab6487186d7739d2"
   Content-Type: application/json
   Server: TornadoServer/4.4.2
   Date: Mon, 20 Feb 2017 12:54:47 GMT

   {"status": "ok", "version": "1.1.1", "name": "my-app-name"}

Python Packaged Application
===========================
If your application is a python package, then you can let the status
handler do the work of looking up your application's name and version
number from the python package metadata.

.. code-block:: python

   from tornado import ioloop, web
   import sprockets_status.handlers

   def make_app(**settings):
       return web.Application([
           web.url('/status', sprockets_status.handlers.StatusHandler,
                   {'package': 'my-app'}),
           # add your handlers here
       ], **settings)

   if __name__ == '__main__':
       app = make_app()
       iol = ioloop.IOLoop.current()
       try:
           app.listen(8888)
           iol.start()
       except KeyboardInterrupt:
           iol.stop()

The status handler will use ``pkg_resources`` to look up the named
distribution and retrieve the package name and version for you.  If you
give it a package that doesn't exist, then it will return a server error
so don't do that.

Developer Quickstart
====================
.. code-block:: bash

   python3.5 -mvenv --copies env
   env/bin/pip install -r requires/development.txt -e .
   env/bin/nosetests
   env/bin/python setup.py build_sphinx

.. _sprockets.handlers.status: https://github.com/sprockets/
   sprockets.handlers.status
.. |Coveralls| image:: https://img.shields.io/coveralls/sprockets/sprockets-status.svg
   :target: https://coveralls.io/github/sprockets/sprockets-status
.. |GitHub| image:: https://img.shields.io/github/release/sprockets/sprockets-status.svg
   :target: https://github.com/sprockets/sprockets-status
.. |PyPI| image:: https://img.shields.io/pypi/v/sprockets-status.svg
   :target: https://pypi.org/project/sprockets-status
.. |Docs| image:: https://img.shields.io/badge/docs-pythonhosted-green.svg
   :target: https://pythonhosted.com/sprockets-status/
.. |Travis| image:: https://img.shields.io/travis/sprockets/sprockets-status.svg
   :target: https://travis-ci.org/sprockets/sprockets-status


docs/history.rst
===============
Release History
===============

`Next Release`_
---------------
- Added :class:`sprockets_status.handlers.StatusHandler`


.. _Next Release: https://github.com/sprockets/sprockets-status/compare/0.0.0...HEAD


tests.py
import json
import pkg_resources

from tornado import testing, web

import examples.app
import sprockets_status.handlers


class SimpleStatusTests(testing.AsyncHTTPTestCase):

    @classmethod
    def setUpClass(cls):
        super(SimpleStatusTests, cls).setUpClass()
        sprockets_status.handlers.StatusHandler._package_name = None

    def get_app(self):
        return examples.app.make_app()

    def test_that_content_type_is_set(self):
        response = self.fetch('/status')
        self.assertEqual(response.code, 200)
        self.assertTrue(
            response.headers['Content-Type'].startswith('application/json'))

    def test_that_application_name_is_included(self):
        response = self.fetch('/status')
        self.assertEqual(response.code, 200)
        body = json.loads(response.body.decode('utf-8'))
        self.assertEqual(body['name'], examples.app.name)

    def test_that_application_version_is_included(self):
        response = self.fetch('/status')
        self.assertEqual(response.code, 200)
        body = json.loads(response.body.decode('utf-8'))
        self.assertEqual(body['version'], examples.app.version)

    def test_that_application_status_is_included(self):
        response = self.fetch('/status')
        self.assertEqual(response.code, 200)
        body = json.loads(response.body.decode('utf-8'))
        self.assertEqual(body['status'], 'ok')


class PackageLookupTests(testing.AsyncHTTPTestCase):

    @classmethod
    def setUpClass(cls):
        super(PackageLookupTests, cls).setUpClass()
        sprockets_status.handlers.StatusHandler._package_name = None

    def setUp(self):
        self.application = None
        super(PackageLookupTests, self).setUp()
        self.package_info = pkg_resources.get_distribution('tornado')

    def get_app(self):
        self.application = web.Application([
            web.url('/status', sprockets_status.handlers.StatusHandler,
                    {'package': 'tornado'}),
        ])
        return self.application

    def test_that_application_name_is_included(self):
        response = self.fetch('/status')
        self.assertEqual(response.code, 200)
        body = json.loads(response.body.decode('utf-8'))
        self.assertEqual(body['name'], self.package_info.project_name)

    def test_that_application_version_is_included(self):
        response = self.fetch('/status')
        self.assertEqual(response.code, 200)
        body = json.loads(response.body.decode('utf-8'))
        self.assertEqual(body['version'], self.package_info.version)

    def test_that_package_lookup_failure_results_in_server_error(self):
        _, spec_list = self.application.handlers[-1]
        spec_list[0].kwargs['package'] = 'not-a-package'
        response = self.fetch('/status')
        self.assertEqual(response.code, 500)

sprockets_status/handlers.py
import json
import logging
import pkg_resources

from tornado import web


_LOGGER = logging.getLogger(__name__)


class StatusHandler(web.RequestHandler):
    """
    Simple handler that returns application status information.

    The easiest way to use this handler is to pass your package's
    name as the ``package`` kwarg to :class:`tornado.web.URLSpec`::

        app = web.Application([
            web.url('/status', StatusHandler, {'package': 'my-package'}),
        ])

    If your application is not bundled as a python package, then you can
    specify the name and version explicitly::

        app = web.Application([
            web.url('/status', StatusHandler,
                    {'name': 'my-package', 'version': '0.0.1'}),
        ])

    """

    OK = 'ok'
    FAILURE = 'failed'

    _package_name = None
    _application_name = 'UNKNOWN'
    _application_version = None

    @property
    def application_name(self):
        """The application's reported name."""
        self._lookup_package_info()
        return self._application_name

    @property
    def application_version(self):
        """The application's version number."""
        self._lookup_package_info()
        return self._application_version

    @property
    def application_status(self):
        """The application's current status."""
        self._lookup_package_info()
        if self.application_name is None:
            return self.FAILURE
        return self.OK

    @classmethod
    def _lookup_package_info(cls):
        if cls._package_name is not None:
            try:
                pkg_info = pkg_resources.get_distribution(cls._package_name)
                cls._application_name = pkg_info.project_name
                cls._application_version = pkg_info.version
                cls._package_name = None
                _LOGGER.debug('found application information "%s" %s',
                              cls._application_name, cls._application_version)
            except pkg_resources.ResolutionError:
                _LOGGER.exception('failed to lookup package "%r"',
                                  cls._package_name)
                cls._application_name = None
                cls._application_version = None

    def initialize(self, **kwargs):
        """
        Sets the static information.

        :keyword str package: name of the package to retrieve
            metadata from
        :keyword str name: name of the application
        :keyword str version: application version number

        If the ``package`` keyword is specified, then the distribution
        is retrieved using :py:mod:`pkg_resources` and used as a source
        of status information.  Otherwise, the ``name`` and ``version``
        keywords are used.

        """
        cls = self.__class__
        if 'package' in kwargs:
            cls._package_name = kwargs['package']
        else:
            cls._application_name = kwargs['name']
            cls._application_version = kwargs['version']

    def get(self):
        """
        Returns a JSON object containing the application status.

        **Sample Response**

        .. code-block:: json

           {
              "name": "application-name",
              "version": "1.2.3",
              "status": "ok"
           }

        """
        if self.application_status == self.FAILURE:
            self.set_status(500)
        else:
            self.set_status(200)

        self.set_header('Content-Type', 'application/json')
        self.write(json.dumps({
            'name': self.application_name,
            'version': self.application_version,
            'status': self.application_status
        }).encode('utf-8'))
        self.finish()
